Pk1d delta extraction

Waelthus opened the pull request:
This is a draft PR for changes to picca_Pk1d and delta_extraction, to re-unify the 1d power analysis with the BAO analyses.
Note that this is not yet complete.

Closes #816 when merged.

Waelthus commented:
now I think most required changes are there, will test next week

Waelthus commented:
This does now work in principle, there are still some missing pieces, e.g. regarding the estimation of noise power based on exposure differences that @corentinravoux worked on. Those will be merged in here as well.

There is a change in data model regarding resolution, in addition to the FWHM-based RESO (in units of km/s) I'm also piping RESO_PIX through the code (incl e.g. masking), i.e. the rms resolution in units of pixel size, which can help for debugging or for building corrections in units different from velocity. I'm also writing both to the outputs.
This required changes in tests as outputs are containing additional info that has not been included before.

Finally, I do the same for the full resolution matrix, if it exists, i.e. delta outputs with pk1d have an additional 11 x npix_out array, which makes them quite a bit larger. It might be useful to make that reso_matrix calculation fully optional and have a config option setting it (so far it is used whenever the inputs have a reso_matrix).

Waelthus commented:
Some of the changes to Pk1d are not pretty (e.g. regarding multiprocessing) and might be changed at a later time. I'll open the PR when an end-to-end test has been successfully done (even for the new computations)

Waelthus commented:
thanks for the comments @iprafols, I'll have a look at them later.

There is one further change that was done in delta_extraction: before we had a cut on SNR when doing continuum fitting with constant weights, this didn't make it into delta_extraction though which messes up the continuum for that case. I did add that back and for the moment activated it for all Pk1d deltas, but it might be that we want to have it for all "use constant weights" analyses instead and we should probably have an option allowing to choose the cut

Waelthus commented on examples/delta_extraction/desi_everest_produce_delta_p1d.ini:
```diff
@@ -0,0 +1,44 @@
+[general]
+overwrite = True
+out dir = $OUTDIR_EVEREST/Delta_lya/
+log = $OUTDIR_EVEREST/Delta_lya/run.log
```
sorry, accidentally pushed that file, this was not finalized yet...

Waelthus commented on py/picca/delta_extraction/astronomical_objects/pk1d_forest.py:
```diff
@@ -163,11 +163,26 @@ def __init__(self, **kwargs):
                                           "Missing variable 'reso'")
         del kwargs["reso"]
 
+        self.reso_pix = kwargs.get("reso_pix")
```
the difference is that `reso` is in units of km/s, whereas reso_pix is in units of pixel. For DESI the latter is roughly constant, leading to the former being roughly propto Delta_log_lambda which is wavelength dependent. For the moment this is more used for debugging purposes, while the actual `resolution_matrix` is what is used in the DESI Pk1d analyses, and the `reso` (or at least its mean) is used for eBOSS

Waelthus commented on py/picca/delta_extraction/astronomical_objects/pk1d_forest.py:
```diff
@@ -163,11 +163,26 @@ def __init__(self, **kwargs):
+                                          "Missing variable 'reso_pix'")
+        del kwargs["reso_pix"]
+
+        self.resolution_matrix = kwargs.get("resolution_matrix")
```
yes, I think we should move everything related to the DESI resolution matrix to DESIPk1dForest, in principle one could call the super().rebin/coadd routines in there and then do the equivalents for just the DESI fields... Just was simpler to implement this way for the moment

iprafols replied:
If you don't mind me pushing directly to this branch, I could move it for you (or I could make a PR to this branch)

Waelthus commented on py/picca/delta_extraction/data_catalogues/desi_tile.py:
```diff
@@ -259,7 +259,7 @@ def read_data(self):
                     flux = spec['FLUX'][w_t].copy()
 
-                    rgs = {
+                    args = {
```
true, can you do this

iprafols replied:
done, see the other PR

Waelthus commented on py/picca/delta_extraction/masks/absorber_mask.py:
```diff
@@ -123,4 +123,7 @@ def apply_mask(self, forest):
             # do the actual masking
             for param in Forest.mask_fields:
-                setattr(forest, param, getattr(forest, param)[w])
+                if param!='resolution_matrix':
```
agreed, better style that way

Waelthus commented on py/picca/pk1d.py:
```diff
@@ -81,31 +90,49 @@ def split_forest(num_parts,
-                  np.power(10., log_lambda_part[0])) / 2. / lambda_abs_igm - 1.0
+
+        if linear_binning:
+            mean_z = np.mean(log_lambda_part) / lambda_abs_igm - 1.0
```
ftm I'm inserting either lambda (and delta_lambda) or log_lambda (and delta_log_lambda) into this function depending on binning. The function assumes linear binning in whatever data is put into it, but I agree the names need to be changed, so that it is clear that it could be either

Waelthus replied:
made function variables more verbose to make clear that either unit could be used

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+    if (q25_lambda - q5_lambda) < 1e-6:
+        #we can assume linear binning for this case
+        linear_binning = True
+        delta_lam = np.median(diff_lambda)
```
you are completely right. the percentile differences are for checking if pixel size changes, but the actual size if that's not the case is given by the minimal size. So I changed this, but in cases with heavy masking (i.e. median!=minimum) the whole Pk1d routine won't work properly anyway.
I tested the routine and indeed it returns the correct binning...

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+    userprint("\n ndata =  ", num_data)
+    results = None
+
+    def process_file(delta):
```
originally I wanted to parallelize over the different deltas in a file, but that requires moving more things around (as the function needs to be definable outside main), atm I'm parallelizing over the whole thing to avoid the complete overhaul (it's not that important yet to parallelize except if running on yr5 predictions) so the function could actually become a for loop again.

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+
+# loop over input files
+def process_all_files(index_file_args):
+    global num_data
```
it's only for counting the number of objects from all different processes together, and while I agree that global vars are ugly, leading to bugs, and should be avoided, in this case it's only used for printing status messages (for which proper process communication would be a little much overhead)

iprafols replied:
We should check where this is defined and used. Currently it is defined as a global variable twice (here and in line 503) and it is only used inside the function `process_all_files` so maybe we could just drop the definition in line 503 and then remove also the `global` keyword

Waelthus replied:
it is only needed such that sub-processes can change the value, it's not needed necessarily except for getting feedback about total progress so far (without the global none of the individual processes could write back to the variable if I remember correctly; or at least not without additional hassle). If you know a good progress-meter option that is multi-processing friendly, or if you'd rather not have progress outputs than inserting a global, I'm personally fine with those options

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
 
-def make_tree(tree, max_num_bins):
-    """Makes the ROOT tree to save the data
+def check_linear_binning(delta):
```
Partly because I didn't ingest how to do this properly, yet. Partly because I forgot about the file that has all config options inside and not only the explicit ones. And partly because in principle this could be run on data created without delta_extraction as well (at least until all the pieces are actually fitting together and e.g. tests have been adapted)...

corentinravoux replied:
I agree that it is better to keep it separate from delta_extraction. I do not know if the delta creation from others will necessarily be implemented in picca.

iprafols replied:
makes sense

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+            return None
+
+        # first pixel in forest
+        selected_pixels = 10**delta.log_lambda > args.lambda_obs_min
```
the filter for both minimal and maximal obs wavelength is actually done in delta_extraction already, this is mostly to avoid e.g. the very noisy blue end of the spectra even if we computed a continuum for it, on the red end there is no such feature (at least not for any reasonable Lya forest redshift range and eBOSS or DESI).

Regarding the other question: at least at the moment the delta object when read from file will always contain `log_lambda` and for this operation the binning is unimportant. The splits later are due to functions that only work on uniform binning (sometimes except for masking), so there binning needs to be tracked more closely. Generally it might be good to convert this routine to a more elegant modular one, but first it needs to work at all (on the new delta outputs)...

iprafols replied:
About the filtering, maybe it would be better to add a comment saying that filtering is already done and that here we just filter more aggressively on the blue end.

Ok I see for the second question. A couple comments on "binning needs to be uniform" should be added to the relevant parts, so that we keep track of it in the future

Waelthus replied:
there are some comments now

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+
+                cols = [k, pk_raw, pk_noise, pk_diff, correction_reso, pk]
+                names = [
+                    'k', 'Pk_raw', 'Pk_noise', 'Pk_diff', 'cor_reso', 'Pk'
```
this was just retained from earlier and I'm not fully sure how dependent on that capitalization my and Corentin's downstream analyses are right now (they should not depend on it, but I'd rather change after testing). Might make unit tests fail as well, but you're right: FITS convention uses all capitals fits column names.

corentinravoux commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+            run_noise = False
+            if args.noise_estimate == 'pipeline':
+                run_noise = True
+            if linear_binning:
+                pk_noise, pk_diff = compute_pk_noise(delta_lambda,
+                                                     ivar_new,
+                                                     exposures_diff_new,
+                                                     run_noise,
+                                                     linear_binning=True)
+            else:
+                pk_noise, pk_diff = compute_pk_noise(delta_log_lambda,
+                                                     ivar_new,
+                                                     exposures_diff_new,
+                                                     run_noise,
+                                                     linear_binning=False)
```
There are a lot of if conditions where only the linear binning changes in this function. Maybe for all of these, you can pass the linear_binning as an argument?

Waelthus replied:
it is an argument for the function already, mostly because things are converted from log-lambda binning to velocity binning internally if the binning is log-lambda. This is not done in case of linear binning.
But also the delta_lambda or delta_log_lambda needs to be put in here depending on binning. This is because here the function needs uniform sampling in whatever binning is used.

Alternatively one could also use the same variables for lambda/log_lambda and delta_lambda/delta_log_lambda and just pass that, but then it's less clear if a variable is in one binning or the other; or one could always pass lambda (or always log_lambda) and then convert the units internally, but that just shifts pain somewhere else...

In the long run I'd guess we need to make the whole computation more modular and properly saving properties like the binning, splitting, filling of pixels, resolution matrix etc in the class doing the computation, but that requires more time

corentinravoux commented on py/picca/delta_extraction/data.py:
```diff
@@ -213,6 +213,13 @@ def filter_forests(self):
                 self.logger.progress(
                     f"Rejected forest with los_id {forest.los_id} "
                     "due to finding nan")
+            elif self.analysis_type=='PK 1D' and forest.mean_snr<1:
```
Maybe I'm wrong but it does not seem that putting a space in a str is a good idea. Why not just 'PK1D'?

Waelthus replied:
this is currently just how it's defined in delta_extraction, not sure at how many places this would need renaming. Also for BAO it is defined as BAO 3d, so it's currently always <statistics> <space> <dimensionality>. I don't really mind doing it either way, but consistency would be most important here...

corentinravoux commented on py/picca/delta_extraction/utils_pk1d.py:
```diff
@@ -16,6 +16,7 @@
 module_logger = logging.getLogger(__name__)
 
+
 def exp_diff(hdul, log_lambda):
```
I do not understand if you merged my 'exp_diff_desi' with the old exp_diff function. If the exp_diff_desi is not added, please add it as it is, and I will make the appropriate changes to adapt it to the data.

Waelthus replied:
no, it's not added yet. The issue is not only the script, but also the interfacing with the data model (iirc currently the diff spectra for desi Pk1d data are just fixed to zero), but we could should change this

Waelthus replied:
the exp_diff_desi is now added as well, needs to be tested, though

corentinravoux reviewed (approved):
With Michael, we started making comparison tests. We will add it to the PR when the plots are done.

iprafols commented:
> thanks for the comments @iprafols, I'll have a look at them later.
> 
> There is one further change that was done in delta_extraction: before we had a cut on SNR when doing continuum fitting with constant weights, this didn't make it into delta_extraction though which messes up the continuum for that case. I did add that back and for the moment activated it for all Pk1d deltas, but it might be that we want to have it for all "use constant weights" analyses instead and we should probably have an option allowing to choose the cut

Waelthus commented:
ok, @corentinravoux: I think the exp_diff_desi should now be usable in principle, is it right that we need to operate on "spectra-..." type files for using it? Or even on some specially made coadds?

I added the option "use non-coadded spectra" to run on spectra files instead of coadd files. Not sure right now if we'd need to change more to allow this mode.

@iprafols: can you check that I introduced the option the right way

iprafols commented on py/picca/delta_extraction/data_catalogues/desi_healpix.py:
```diff
@@ -62,6 +69,10 @@ def __init__(self, config):
 
+        self.use_non_coadded_spectra = config.getboolean("use non-coadded spectra")
+        if self.use_single_nights is None:
+            raise DataError("Missing argument 'use non-coadded spectra' required by DesiTile")
```
This should be `DesiHealpix` and not `DesiTile`

iprafols commented on py/picca/delta_extraction/data_catalogues/desi_healpix.py:
```diff
@@ -62,6 +69,10 @@ def __init__(self, config):
         """
         self.logger = logging.getLogger(__name__)
 
+        self.use_non_coadded_spectra = config.getboolean("use non-coadded spectra")
```
Usually, I add this in a `_parse_config` method (as in DesiTile, for example), it might be good to add it here too in case we add more options in the future

Waelthus replied:
note that this addition was less straight-forward than expected, as when I first had the method included, the `_parse_config` of the parent class was silently ignored without tests failing (but failing at runtime due to eg missing input_directory)

iprafols commented on py/picca/delta_extraction/data_catalogues/desi_tile.py:
```diff
@@ -102,6 +103,10 @@ def _parse_config(self, config):
         if self.use_single_nights is None:
             raise DataError("Missing argument 'use single nights' required by DesiTile")
 
+        self.use_non_coadded_spectra = config.getboolean("use non-coadded spectra")
```
You should insert `self.use_single_nights = None` between lines 82 and 83

Waelthus commented:
@iprafols there's something weird with the rebinning in here, to make tests pass I needed to change `DesiForest.__init__` to end with `self.rebin()` instead of `super().rebin()`. Else some quantities (e.g. flux) would end up rebinned after init, some others (e.g. resoltion_matrix) wouldn't (calling rebin later on the child classes does not work because once ivar has been rebinned all indices are out of sync). 
Not rebinning in `DesiForest.__init__` does not seem to work either (even though the last `super().__init__()` should end with a `self.rebin` by `Forest`).
I don't think I'm properly getting how the rebinning at initialization should work (I'm not actually sure why we do it at that step and not as one additional rebin operation after all data has been initialized).
Can you elaborate a bit? This might also be happening partly because `DesiPk1dForest` inherits both from `Pk1dForest` and `DesiForest`.

iprafols commented:
Hi @Waelthus , 
It seems it is likely to be a problem with the inheritance pattern. You are right that it makes life simpler if forests are rebinned after initialisation (and not during). It makes the whole inheritance flow easier. I'll do a PR to fix this

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,367 @@
+                    'comment': "analysis was performed on delta with linear binned lambda"
+                }, {
+                    'name': 'LOS_ID',
+                    'value': (delta.targetid if hasattr(delta, "targetid")
```
I don't think it's in the output (or not read by data.delta.from_fits), but I'll check again.

Waelthus commented:
I needed to backport the summation over multiple exp in one file to allow reading in the spectra for exposure diff calculations, so this now also adds a way of simply coadding the spectra.

Waelthus commented on py/picca/data.py:
```diff
@@ -1068,16 +1070,28 @@ def from_fitsio(cls, hdu, pk1d_type=False):
             mean_z = header['MEANZ']
-            delta_log_lambda = header['DLL']
+            #this is neither stored nor really needed
+            #delta_log_lambda = header['DLL']
```
removed it

Waelthus commented:
ok, this branch now runs on healpix data and mocks, and passes through resolution etc, some testing of the results is still missing

Waelthus commented:
This shows the current differences between picca_deltas + picca_pk1d from old branch and the new picca_delta_extraction + updated picca_pk1d from this branch.
Overall this seems to be ok, e.g. the most complicated change regarding pass-through of the resolution matrix and correction based on that matrix seem to work the same as previously. But we still need to figure out a difference in datasets, in the end-to-end comparison we currently have 10-20% more chunks (which are 1/3 of a spectrum each) left.

![image](https://user-images.githubusercontent.com/3012942/157428442-a8e56bc9-0b79-4395-87c0-3ecfcbc122c4.png)

Waelthus commented:
Residual differences in power spectrum estimates, on large scales dominated by mild continuum mismatch (one pixel more in the new mean cont than the old one, mild scatter in pixel size for the old mean cont), on small scales by randomness in the noise power estimate
![image](https://user-images.githubusercontent.com/3012942/158557411-2a4adc0d-39d1-4440-a8cd-d018f52817fb.png)

![image (1)](https://user-images.githubusercontent.com/3012942/158557938-7353f771-2b2b-46a5-a2b2-06f2f63251f6.png)

Waelthus commented:
After the fix rebin changes

![cont](https://user-images.githubusercontent.com/3012942/159237576-53ce9782-cc9f-49ed-9798-c6e3b6e5125c.png)
![pk1d](https://user-images.githubusercontent.com/3012942/159237578-2dc1b567-4a64-4cfb-ac6d-aebe3cf8736e.png)

iprafols commented on py/picca/delta_extraction/astronomical_objects/desi_pk1d_forest.py:
```diff
@@ -145,17 +123,158 @@ class DesiPk1dForest(DesiForest, Pk1dForest):
+
+        Raise
+        -----
+        AstronomicalObjectError if other is not a Pk1dForest instance
```
Docstring should say `DesiPk1dForest` instead of `Pk1dForest`

iprafols commented on py/picca/delta_extraction/data_catalogues/desi_healpix.py:
```diff
@@ -83,10 +116,11 @@ def read_data(self):
+            healpy.ang2pix(in_nside,
+                           np.pi / 2 - row["DEC"],
+                           row["RA"],
+                           nest=True) for row in self.catalogue
```
Maybe the `for` bit can be in another line aligned with `healpy.ang2pix...`

iprafols reviewed (changes requested):
The changes look good, but I have a few suggestions to improve the code

Waelthus commented on py/picca/delta_extraction/astronomical_objects/desi_forest.py:
```diff
@@ -19,48 +19,22 @@ class DesiForest(Forest):
-    log_lambda_min_rest_frame: float or None (from Forest)
-    As log_lambda_min but for rest-frame wavelength. This should not be None if
-    wave_solution is "log". Ignored if wave_solution is "lin".
+    blinding: str (from Forest)
```
I guess in that case it might be less of a fuss if we merge #851 first

Waelthus commented on py/picca/delta_extraction/data_catalogues/desi_healpix.py:
```diff
@@ -163,14 +204,33 @@ def read_file(self, filename, catalogue, forests_by_targetid):
+                    else:
+                        spec['TEFF_LYA'] = np.ones(spec["FLUX"].shape[0])
+                        if self.use_non_coadded_spectra and not no_scores_available:
+                            self.logger.info("SCORES are missing, Teff information (and thus DIFF) will be garbage")
```
I guess we can by now expect that this is always available, warning makes sense though. Running can still be useful e.g. when wanting to compare different coadd scenarios without actually needing the difff, most analysis use the pipeline noise anyway.

Waelthus commented on py/picca/delta_extraction/data_catalogues/desi_healpix.py:
```diff
@@ -163,14 +204,33 @@ def read_file(self, filename, catalogue, forests_by_targetid):
-                            "{filename}. Analysis type is  'PK 1D', "
-                            "but file does not contain HDU "
-                            f"'{color}_RESOLUTION' ")
+                        basename_truth=os.path.basename(filename).replace('spectra-','truth-')
```
no, actually this is a thing only needed for mocks where the resolution is in a truth file. So I'll raise the dataerror if there is no truth file as for data. and read the truth file for a model. Else we'd probably need to split up the i/o functions further such that we can load only relevant parts in the data_healpix vs desisim_mocks (that currently just inherits the function from here...)

Waelthus commented:
Thanks @iprafols for your comments. I did address everything that is relevant for delta extraction. I'll need to think about the comments regarding pk1d. I agree that some of the things are written in a very ugly way, but atm they also agree with the behaviour of previous versions...
Given that changes in pk1d do not involve further chanegs inside delta_extraciton routines I'd suggest merging this anyway and fixing the pk1d naming/documentation in a continuation PR. That allows further development on delta_extraction without accidentally breaking the pk1d part and allows running pk1d without referring to special branches.

Waelthus commented on py/picca/pk1d.py:
```diff
@@ -12,28 +12,32 @@
-                 delta_log_lambda,
-                 log_lambda,
+                 delta_lambda_or_log_lambda,
+                 lambda_or_log_lambda,
```
let's do larger re-writes in a seperate

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,361 @@
+
+# loop over input files
+def process_all_files(index_file_args):
+    global num_data  #this only stores the number of spectra for progress viewing purposes
```
I think it will only count per file then. But I guess it's ok anyway... And removes the global

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,361 @@
+        num_parts = min(args.nb_part, max_num_parts)
+
+        #the split_forest function works with either binning, but needs to be uniform
+        if linear_binning:
```
indeed that's right. And it would make sense, but lets do further cleanup in a seperate PR

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -7,152 +7,361 @@
+        selected_pixels = 10**delta.log_lambda > args.lambda_obs_min
+        #this works as selected_pixels returns a bool and argmax points
+        #towards the first occurance for equal values
+        first_pixel_index = (np.argmax(selected_pixels) if
```
I'm keeping it as is for the moment, it's all legacy things that are the same in the current master anyways, but will reconsider in a later pr

Waelthus commented:
Did some changes for better readability, but not all. Think this is good to go?

iprafols reviewed (approved):
Things look fine now! Thanks for implementing these changes. I agree we can leave the changes outside delta_extraction for another PR. @Waelthus feel free to merge whenever you are ready to do so
